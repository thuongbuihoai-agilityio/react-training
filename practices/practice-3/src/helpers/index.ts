const setFieldsValue = (fieldsValue: any, value: string, fieldName: string) => {
  const fieldData = fieldsValue[fieldName];
  return {
    ...fieldsValue,
    [fieldName]: {
      ...fieldData,
      value: value,
      error: ""
    }
  };
}

export { setFieldsValue };
