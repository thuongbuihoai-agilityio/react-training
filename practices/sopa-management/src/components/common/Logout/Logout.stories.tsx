import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Logout from '@common/Logout';

const meta: Meta<typeof Logout> = {
  component: Logout,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    )
  ]
};

export default meta;
type Story = StoryObj<typeof Logout>;

export const LogoutDefault: Story = {
  args: {}
};
