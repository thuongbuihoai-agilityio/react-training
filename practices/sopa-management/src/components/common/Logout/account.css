.account {
  position: relative;
}

.account-dropdown {
  position: absolute;
}