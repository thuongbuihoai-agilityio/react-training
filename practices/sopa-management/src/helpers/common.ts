// Interfaces
import { Product } from '@interfaces/product';

// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};

// Calculate the total price
export const totalPrices = (carts: Product[]) => {
  const subTotals = carts.map((item: Product) => item.quantity * item.price);

  const total = subTotals.reduce(
    (totalPrice: number, cartItem: number) => totalPrice + cartItem,
    0
  );

  return total;
};
