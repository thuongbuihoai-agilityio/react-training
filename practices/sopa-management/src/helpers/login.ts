import {
  Account,
  CheckType,
  LoginType
} from "@interfaces/account";
import { FieldError } from "react-hook-form";

// Check login
export const checkAccount = (
  data?: Account[],
  email?: string,
  password?: string,
  checkType?: CheckType,
) => {
  return data?.some((user: Account) => {
    const checkEmail = user.email === email;
    const checkPassword = user.password === password;
    const checkLogin = checkType === CheckType.login && checkEmail && checkPassword;

    if (checkLogin
      || checkType === CheckType.email && checkEmail
      || checkType === CheckType.password && checkPassword
    ) {
      return true;
    }
    return false;
  })
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
  type: LoginType,
) => {
  const checkIsDirty = isDirty ? type.info : type.default
  const style = serverError || rules ? type.error : checkIsDirty;

  return { style };
};
