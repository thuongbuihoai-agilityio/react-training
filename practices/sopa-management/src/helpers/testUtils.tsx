import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
