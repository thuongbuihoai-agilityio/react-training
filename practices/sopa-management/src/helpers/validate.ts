import { REGEX } from "@constants/regex"
import { ERROR_MESSAGES } from "@constants/validate"

// Validate
export const VALIDATE = {
  email: {
    required: ERROR_MESSAGES.REQUIRED,
    pattern: {
      value: REGEX.EMAIL,
      message: ERROR_MESSAGES.EMAIL_FORMAT,
    },
  },
  password: {
    required: ERROR_MESSAGES.REQUIRED,
    pattern: {
      value: REGEX.PASSWORD,
      message: ERROR_MESSAGES.PASSWORD_FORMAT,
    },
  },
}
