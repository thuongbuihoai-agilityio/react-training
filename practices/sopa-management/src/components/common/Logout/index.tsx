import {
  memo,
  useCallback,
  useState
} from 'react';
import { useNavigate } from 'react-router-dom';
import { shallow } from 'zustand/shallow';

// Components
import {
  Button,
  ButtonType,
  Dropdown,
  Icon,
  IconType
} from '@components/common';

// Constants
import { useAuthenticationStore } from '@stores/login';

interface LogoutProps {
  className?: string;
}

const Logout = ({ className = '' }: LogoutProps) => {
  const navigate = useNavigate();

  const [logout] = useAuthenticationStore(
    (state) => [
      state.logout,
    ],
    shallow
  );

  const [showDropdown, setShowDropdown] = useState<boolean>(false);

  const handleToggle = useCallback(() => {
    setShowDropdown(!showDropdown);
  }, [showDropdown]);

  const handleLogout = () => {
    useAuthenticationStore.persist.clearStorage();
    logout();
    navigate('/login');
  };

  const handleToggleItem = useCallback(() => {
    handleToggle();
    handleLogout();
  }, [showDropdown]);

  return (
    <div className={`${className} account`} data-testid='logout'>
      <Button
        ariaLabel='User'
        children={<Icon iconName={IconType.user} />}
        type={ButtonType.btnIconPrimary}
        onClick={handleToggle}
      />
      {showDropdown && (
        <Dropdown
          isHref
          className='account-dropdown'
          onClick={handleToggleItem}
        />
      )}
    </div>
  );
};

export default memo(Logout);
