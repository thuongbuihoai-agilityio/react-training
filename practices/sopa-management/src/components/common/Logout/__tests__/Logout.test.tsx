// Libs
import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import Logout from '@common/Logout';

// Helpers
import { renderRouterTest } from '../../../../helpers/testUtils';

describe('Logout component', () => {
  test('should render Logout component', () => {
    const { getByTestId } = renderRouterTest(<Logout />);
    expect(getByTestId('logout')).toBeInTheDocument();
  });

  test('Should render Logout component with action onToggle', async () => {
    const { getByTestId } = renderRouterTest(<Logout />);

    const actionButton = getByTestId('button');
    await userEvent.click(actionButton);

    const logoutButton = getByTestId('dropdown');
    await userEvent.click(logoutButton);

    await waitFor(() => expect(logoutButton).toBeInTheDocument());
  });

  test('matches snapshot', () => {
    const { asFragment } = renderRouterTest(<Logout />);
    expect(asFragment()).toMatchSnapshot();
  });
});
