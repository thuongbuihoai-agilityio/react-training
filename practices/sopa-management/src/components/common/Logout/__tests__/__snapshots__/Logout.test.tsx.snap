// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=" account"
    data-testid="logout"
  >
    <button
      aria-label="User"
      class=" btn btn-icon-primary"
      data-testid="button"
    >
      <div
        class="undefined icon-user"
        data-testid="icon"
      />
    </button>
  </div>
</DocumentFragment>
`;
